File: 2018/day14.js
import fs from "fs";
import path from "node:path";

console.time();

let inputStr = fs.readFileSync(
  path.join(import.meta.dirname, "inputs", "day14.txt"),
  "utf-8",
);
let input = parseInt(inputStr);
let inputDigits = inputStr.split('').map(digit => parseInt(digit));

let idx1 = 0;
let idx2 = 1;
let recipes = [3, 7];
while (recipes.length < input + 10) {
  createNewRecipes(recipes[idx1], recipes[idx2]);
  moveElves();
}
console.log(`Part 1: ${recipes.slice(input, input + 10).join("")}`);

while (true) {
  let match = createNewRecipes(recipes[idx1], recipes[idx2]);
  if (match) {
    console.log(`Part 2: ${recipes.length - inputDigits.length}`);
    break;
  }
  moveElves();
}

console.timeEnd();

function createNewRecipes(score1, score2) {
  let sum = score1 + score2;
  if (sum < 10) {
    recipes.push(sum);
    if (recipesEndWithInput()) {
      return true;
    }
  } else {
    recipes.push(1);
    if (recipesEndWithInput()) {
      return true;
    }
    recipes.push(sum - 10);
    if (recipesEndWithInput()) {
      return true;
    }
  }
  return false;
}

function moveElves() {
  idx1 = (idx1 + recipes[idx1] + 1) % recipes.length;
  idx2 = (idx2 + recipes[idx2] + 1) % recipes.length;
}

function recipesEndWithInput() {
  for (let i = 0; i < inputDigits.length; i++) {
    if (recipes.at(-inputDigits.length + i) !== inputDigits[i]) {
      return false
    }
  }
  return true;
}

File: 2018/day14.py
import time

def main():
    input_str = "640441"
    target = int(input_str)
    input_digits = [int(digit) for digit in list(input_str)]

    idx_1 = 0
    idx_2 = 1
    recipes = [3, 7]

    while len(recipes) < target + 10:
        create_new_recipes(recipes, recipes[idx_1], recipes[idx_2], input_digits)
        new_idxs = move_elves(recipes, idx_1, idx_2)
        idx_1 = new_idxs[0]
        idx_2 = new_idxs[1]
    print(f"Part 1: {''.join([str(elem) for elem in recipes[target:target+10]])}")
    
    while True:
        match = create_new_recipes(recipes, recipes[idx_1], recipes[idx_2], input_digits)
        if match:
            print(f"Part 2: {len(recipes) - len(input_digits)}")
            break
        new_idxs = move_elves(recipes, idx_1, idx_2)
        idx_1 = new_idxs[0]
        idx_2 = new_idxs[1]

def create_new_recipes(recipes, score_1, score_2, input_digits):
    sum = score_1 + score_2
    if sum < 10:
        recipes.append(sum)
        if recipes_end_with_input(recipes, input_digits):
            return True
    else:
        recipes.append(1)
        if recipes_end_with_input(recipes, input_digits):
            return True
        recipes.append(sum - 10)
        if recipes_end_with_input(recipes, input_digits):
            return True
    return False

def move_elves(recipes, idx_1, idx_2):
    return (
        (idx_1 + recipes[idx_1] + 1) % len(recipes),
        (idx_2 + recipes[idx_2] + 1) % len(recipes)
    )

def recipes_end_with_input(recipes, input_digits):
    if len(recipes) < len(input_digits): return False
    for i in range(0, len(input_digits)):
        if recipes[-len(input_digits) + i] != input_digits[i]:
            return False
    return True

if __name__ == "__main__":
    start_time = time.time()
    main()
    print("--- %s seconds ---" % (time.time() - start_time))
